src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/config/UserVersionFilter.kt

```diff
@@ -0,0 +1,47 @@
+package uk.gov.justice.digital.hmpps.approvedpremisesapi.config
+
+import org.springframework.stereotype.Component
+import org.springframework.web.filter.OncePerRequestFilter
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.api.model.ServiceName
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.UserRole
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.service.RequestContextService
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.service.SentryService
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.service.UserService
+import java.io.IOException
+import javax.annotation.PostConstruct
+import javax.servlet.FilterChain
+import javax.servlet.ServletException
+import javax.servlet.http.HttpServletRequest
+import javax.servlet.http.HttpServletResponse
+
+@Component
+class UserVersionFilter(
+  val sentryService: SentryService,
+  private val userService: UserService,
+  private val requestContextService: RequestContextService,
+) : OncePerRequestFilter() {
+
+  @PostConstruct
+  fun filterStartup() {
+    sentryService.captureErrorMessage("User Version filtering is enabled.")
+  }
+
+  @Throws(ServletException::class, IOException::class)
+  override fun doFilterInternal(request: HttpServletRequest, response: HttpServletResponse, filterChain: FilterChain) {
+    when (requestContextService.getServiceForRequest()) {
+      ServiceName.approvedPremises -> {
+        addUserVersionHeader(response)
+        filterChain.doFilter(request, response)
+      }
+      else -> {
+        return filterChain.doFilter(request, response)
```

Comment:
After running this locally, i think the header wants setting _after_ the filterChain.doFilter(request,response). Then it will reflect any changes to the user made in this request (e.g. if user is admin and change their own permissions), This is really handy for demos

Reply:
Moving the ordering resulted in the header not being present in the response.   As discussed maybe this is something to tackle later, a custom filter strategy to add the header last?